CMakeLists.txt

```diff
@@ -161,10 +161,25 @@ ELSE()
   ENDIF()
   ADD_DEBUG_COMPILE_FLAG("-O0")
   ADD_DEBUG_COMPILE_FLAG("-g3")
-  ADD_NONDEBUG_COMPILE_FLAG("-O2")
+  IF (EMSCRIPTEN)
+    # really focus on minimizing output size when compiling sources
+    ADD_NONDEBUG_COMPILE_FLAG("-Oz")
+  ELSE()
+    ADD_NONDEBUG_COMPILE_FLAG("-O2")
+  ENDIF()
   ADD_NONDEBUG_COMPILE_FLAG("-UNDEBUG") # Keep asserts.
 ENDIF()
 
+IF (EMSCRIPTEN)
+  # link with -O3 for metadce and other powerful optimizations. note that we
+  # must use add_link_options so that this appears after CMake's default -O2
+  add_link_options("-O3")
+  ADD_LINK_FLAG("-s SINGLE_FILE")
```

Comment:
What's the difference between `add_link_options` and `ADD_LINK_FLAG`?

Reply:
In practice, `add_link_options` is the only way that I can find to append a "final" flag. cmake also emits a `-O2` by default for an opt build, but we want `-O3` in emscripten.

`add_link_options` seems like the normal way to do things https://cmake.org/cmake/help/git-stage/command/add_link_options.html while `ADD_LINK_FLAG` is something that was added in this cmake script by us, which does this:
```cmake
FUNCTION(ADD_LINK_FLAG value)
  MESSAGE(STATUS "Linking with ${value}")
  FOREACH(variable CMAKE_EXE_LINKER_FLAGS)
    SET(${variable} "${${variable}} ${value}" PARENT_SCOPE)
  ENDFOREACH(variable)
ENDFUNCTION()
```
Not sure how all that works, but it ends up before cmake's own `-O2` which is bad...